htlcswitch/link.go

```diff
@@ -1257,23 +1279,26 @@ func (l *channelLink) htlcManager() {
 				)
 			}
 
-		case req := <-l.shutdownRequest:
-			// If the channel is clean, we send nil on the err chan
-			// and return to prevent the htlcManager goroutine from
-			// processing any more updates. The full link shutdown
-			// will be triggered by RemoveLink in the peer.
-			if l.channel.IsChannelClean() {
-				req.err <- nil
-				return
-			}
-
-			// Otherwise, the channel has lingering updates, send
-			// an error and continue.
-			req.err <- ErrLinkFailedShutdown
+		case <-l.shutdownRequest:
+			return
 
 		case <-l.quit:
 			return
 		}
+
+		// After we are finished processing the event, if the link is
+		// flushing, we check if the channel is clean and invoke the
+		// post-flush hook if it is.
+		if l.IsFlushing() && l.channel.IsChannelClean() {
```

Comment:
We can only shutdown here if we've also received their `Shutdown`. `IsFlushing` may be set when we're the initiator meaning we haven't received their `Shutdown`. The counterparty is still allowed to add HTLC's in this case according to the spec

Reply:
I think what we'll need is one variable to know when we've received their `Shutdown` and another variable to know when we've sent our `Shutdown`

Reply:
I'll take a closer look at the spec and see what we can do. I was trying to avoid duplicating the ChanCloser state logic inside of the link since that was a barrier in the PR that precedes this one. As far as I know the ChanCloser is responsible for the state machine from the "initial state" where no `shutdown`s have been sent all the way through `closing_signed`s. Is there a reason you think this has to be duplicated in the link itself? Also keep in mind that if we had a "flush" operation that wasn't a shutdown (dyncomm execution), then this wouldn't translate properly. Conceptually all this does is execute the continuation at the first clean state after the flush has been initiated. The shutdown use case puts the shutdown logic into that continuation.

Reply:
Currently flushing means that `Flush` has been called once (either we've sent `Shutdown` or the peer has), but I think there could be a new interface call that means we've sent and received `Shutdown`. The issue is that if we've sent `Shutdown`, the peer can still send HTLCs or `update_fee` and if we've exited, the peer will still be waiting for that HTLC or `update_fee` to get resolved. I'm not sure whether this translates well to dynamic commitments, but ideally we'd find a way to do that as well

Reply:
So the `Flush` can only begin once we've both received and sent `shutdown`, then.